zqd/handlers.go

```diff
@@ -0,0 +1,217 @@
+	}
+	s, err := space.Open(root, req.Space)
+	if err != nil {
+		status := http.StatusInternalServerError
+		if err == space.ErrSpaceNotExist {
+			status = http.StatusNotFound
+		}
+		http.Error(w, err.Error(), status)
+		return
+	}
+	var out search.Output
+	format := r.URL.Query().Get("format")
+	switch format {
+	case "zjson", "json":
+		// XXX Should write appropriate ndjson content header.
+		out = search.NewJSONOutput(w, search.DefaultMTU)
+	case "bzng":
+		// XXX Should write appropriate bzng content header.
+		out = search.NewBzngOutput(w)
+	default:
+		http.Error(w, fmt.Sprintf("unsupported output format: %s", format), http.StatusBadRequest)
+	}
+	// XXX This always returns bad request but should return status codes
+	// that reflect the nature of the returned error.
+	if err := search.Search(r.Context(), s, req, out); err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+}
+
+func handlePacketSearch(root string, w http.ResponseWriter, r *http.Request) {
+	spaceName, err := extractSpace(r)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	s, err := space.Open(root, spaceName)
+	if err != nil {
+		status := http.StatusInternalServerError
+		if err == space.ErrSpaceNotExist {
+			status = http.StatusNotFound
+		}
+		http.Error(w, err.Error(), status)
+		return
+	}
+	req := &api.PacketSearch{}
+	if err := req.FromQuery(r.URL.Query()); err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+	}
+	if !s.HasFile(pcapFile) || !s.HasFile(pktIndexFile) {
+		http.Error(w, "space has no pcaps", http.StatusNotFound)
+		return
+	}
+	index, err := pcap.LoadIndex(s.DataPath(pktIndexFile))
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	// XXX pcapFile should be able to live outside space data path?
+	f, err := s.OpenFile(pcapFile)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	defer f.Close()
+	slicer, err := pcap.NewSlicer(f, index, req.Span)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	search, err := pcap.NewSearch(
+		req.Span,
+		req.Proto,
+		req.SrcHost,
+		req.SrcPort,
+		req.DstHost,
+		req.DstPort,
+	)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	w.Header().Set("Content-Type", "application/vnd.tcpdump.pcap")
+	w.Header().Set("Content-Disposition", fmt.Sprintf("inline; filename=%s.pcap", search.ID()))
+	err = search.Run(w, slicer)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+	}
+}
+
+func handleSpaceList(root string, w http.ResponseWriter, r *http.Request) {
+	info, err := ioutil.ReadDir(root)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	spaces := []string{}
+	for _, subdir := range info {
+		if !subdir.IsDir() {
+			continue
+		}
+		s, err := space.Open(root, subdir.Name())
+		if err != nil || s.HasFile("all.bzng") {
```

Comment:
Should this be `!s.HasFile`?

Reply:
Whoops. I suppose if I had a test of space list this would have been caught...

Reply:
Actually I'm going to get rid of the check for `all.bzng`. As space still exists if it can open (i.e. has a valid `config.json`) but has no data.